### ISSUE TITLE
Responsive images (derivatives) generation on image upload or via option in the admin backend

### ISSUE THREAD

**Issue by HoloBack** [getgrav/grav#2918]
Currently, in the latest version of GRAV the image derivatives are generated on the first page request, which can take a lot of processing time. Are you working on a feature to generate the images on upload (f.e. according to admin settings) or through a button "Generate responsive images" in the backend? This would be a very important feature.

I tried some plugins which should solve this problem, but none of them worked for me.

**Comment by ricardo118** [comment]
@HoloBack if you hit preview in admin, would that technically do first page request? essentially generating the images?

**Comment by HoloBack** [comment]
@ricardo118 Thanks for the fast reply! Yes it does the first page request, but it doesn't solve the problem - you just click through all the pages manually in preview mode. I think it would save a lot of time, if images would be generated on upload and additionally with a button "Generate images", where a background process is started which iterates through every page and generates the images. That's how we all know it from Wordpress and other CMS.

**Comment by arkhi** [comment]
I would love this feature too. Generating derivatives for each image on upload would be less intensive and hopefully avoid server timeout in the case of a lot of images with derivatives on one page.

**Comment by kali-hernandez** [comment]
How does it behave in this first page request scenario, if there are concurrent first page requests? Would it trigger a race condition? 
That would be a much more important reason to have the generation run in the background on upload rather than on page display

**Comment by HoloBack** [comment]
@arkhi You're right, currently we are using the Warm Cache plugin which does a good job, but sometimes crashes with a 500 when processing to many images. I think processing them on upload like f.e. Wordpress does, is a good approach.

Good point @kali-hernandez, that's an interesting scenario too.
